src/mc_tasks/lipm_stabilizer/StabilizerTask.cpp

```diff
@@ -574,17 +604,89 @@ void StabilizerTask::staticTarget(const Eigen::Vector3d & com, double zmpHeight)
 void StabilizerTask::target(const Eigen::Vector3d & com,
                             const Eigen::Vector3d & comd,
                             const Eigen::Vector3d & comdd,
-                            const Eigen::Vector3d & zmp)
+                            const Eigen::Vector3d & zmp,
+                            const Eigen::Vector3d & zmpd)
 {
-  comTarget_ = com;
+  comTargetRaw_ = com;
+  comTarget_ = comTargetRaw_ - comOffsetErrCoM_;
+  if(c_.extWrench.addExpectedCoMOffset)
+  {
+    comTarget_ -= comOffsetTarget_;
+  }
   comdTarget_ = comd;
   comddTarget_ = comdd;
   zmpTarget_ = zmp;
+  zmpdTarget_ = zmpd;
   double comHeight = comTarget_.z() - zmpTarget_.z();
   omega_ = std::sqrt(constants::gravity.z() / comHeight);
   dcmTarget_ = comTarget_ + comdTarget_ / omega_;
 }
 
+void StabilizerTask::setExtWrenches(const std::vector<std::pair<std::string, sva::ForceVecd>> & extWrenches)
+{
+  extWrenchesTarget_ = extWrenches;
+  comOffsetTarget_ = computeCoMOffset(extWrenchesTarget_, robot());
+  comTarget_ = comTargetRaw_ - comOffsetErrCoM_;
+  if(c_.extWrench.addExpectedCoMOffset)
+  {
+    comTarget_ -= comOffsetTarget_;
+  }
+}
+
+Eigen::Vector3d StabilizerTask::computeCoMOffset(const std::vector<std::pair<std::string, sva::ForceVecd>> & extWrenches,
+                                                 const mc_rbdyn::Robot & robot) const
+{
+  Eigen::Vector3d comOffset = Eigen::Vector3d::Zero();
+
+  for(const auto & extWrench : extWrenches)
+  {
+    Eigen::Vector3d pos, force, moment;
+    computeExtContact(robot, extWrench.first, extWrench.second, pos, force, moment);
+
+    comOffset.x() += (pos.z() - zmpTarget_.z()) * force.x() - pos.x() * force.z() + moment.y();
+    comOffset.y() += (pos.z() - zmpTarget_.z()) * force.y() - pos.y() * force.z() - moment.x();
+  }
+  comOffset /= robot.mass() * (comddTarget_.z() + constants::gravity.z());
```

Comment:
Potential division by zero if `comddTarget_.z() = -constants::gravity.z()`.

Reply:
Fixed in https://github.com/mmurooka/mc_rtc/commit/f8c8bd48677b0f5cdfc1aff44c213aeb35c5a6e7. (Please update it if there is a better way.)

Since `constants::gravity.z() = 9.8` and `comddTarget_.z()` is usually around zero, it should be positive in normal motion.